include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -85,172 +130,291 @@ namespace internal
+      apply<direction,contract_over_rows,add,true>(shape_gradients, in, out);
+    }
+
+    template <int direction, bool contract_over_rows, bool add>
+    void
+    hessians_one_line (const Number in [],
+                       Number       out[]) const
+    {
+      Assert (shape_hessians != nullptr, ExcNotInitialized());
+      apply<direction,contract_over_rows,add,true>(shape_hessians, in, out);
     }
 
-    template <int direction, bool dof_to_quad, bool add>
-    static void apply (const Number *shape_data,
-                       const Number in [],
-                       Number       out []);
+    /**
+     * This function applies the tensor product kernel, corresponding to a
+     * multiplication of 1D stripes, along the given @p direction of the tensor
+     * data in the input array. This function allows the @p in and @p out
+     * arrays to alias for the case n_rows == n_columns, i.e., it is safe to
+     * perform the contraction in place where @p in and @p out point to the
+     * same address. For the case n_rows != n_columns, the output is in general
+     * not correct.
+     *
+     * @tparam direction Direction that is evaluated
+     * @tparam contract_over_rows If true, the tensor contraction goes sums
+     *                            over the rows in the given @p shape_data
+     *                            array, otherwise it sums over the columns
+     * @tparam add If true, the result is added to the output vector, else
+     *             the computed values overwrite the content in the output
+     * @tparam one_line If true, the kernel is only applied along a single 1D
+     *                  stripe within a dim-dimensional tensor, not the full
+     *                  n_rows^dim points as in the @p false case.
+     *
+     * @param shape_data Transformation matrix with @p n_rows rows and
+     *                   @p n_columns columns, stored in row-major format
+     * @param in Pointer to the start of the input data vector
+     * @param out Pointer to the start of the output data vector
+     */
+    template <int direction, bool contract_over_rows, bool add, bool one_line=false>
+    static void apply (const Number2 *DEAL_II_RESTRICT shape_data,
+                       const Number  *in,
+                       Number        *out);
 
-    const Number *shape_values;
-    const Number *shape_gradients;
-    const Number *shape_hessians;
+    /**
+     * This function applies the tensor product operation to produce face values
+     * from cell values. As opposed to the apply method, this method assumes
+     * that the directions orthogonal to the face have n_rows degrees of
+     * freedom per direction and not n_columns for those directions lower than
+     * the one currently applied. In other words, apply_face() must be called
+     * before calling any interpolation within the face.
+     *
+     * @tparam direction Direction of the normal vector (0=x, 1=y, etc)
```

Comment:
is it better to call it `face_direction`?

Reply:
Fixed. (This was a typo in the comment, the actual parameter was already there.)